designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,784 @@
+1. There is no automatic merging of config files using either `extends` or `overrides`.
+1. When `-c` is passed on the command line, there is no search performed.
+
+### File Pattern Resolution
+
+Because there are file patterns included in `eslint.config.js`, this requires a change to how ESLint decides which files to lint. The process for determining which files to lint is:
+
+1. When a filename is passed directly (such as `eslint foo.js`):
+    1. ESLint checks to see if there is one or more configs where the `files` pattern matches the file that was passed in and does not match the `ignores` pattern. The pattern is evaluated by prepending the directory in which `eslint.config.js` was found to each pattern in `files`. All configs that match `files` and not `ignores` are merged (with the last matching config taking precedence over others). If no config is found, then the file is ignored with an appropriate warning.
+    1. If a matching config is found, then the `ignores` pattern is tested against the filename. If it's a match, then the file is ignored. Otherwise, the file is linted.
+1. When a glob pattern is passed directly (such as `eslint src/*.js`):
+    1. ESLint expands the glob pattern to get a list of files.
+    1. Each file is checked individually as in step 1.
+1. When a directory is passed directly (such as `eslint src`):
+    1. The directory is converted into a glob pattern by appending `**/*` to the directory (such as `src` becomes `src/**/*`).
+    1. The glob pattern is checked as in step 2.
+1. When a relative directory is passed directly (such as `eslint .`):
+    1. The relative directory pattern is resolved to a full directory name.
+    1. The glob pattern is checked as in step 3.
+
+**Note:** ESLint will continue to ignore `node_modules` by default.
+
+### Rename `--no-eslintrc` to `--no-config-file` and `useEslintrc` to `useConfigFile`
+
+Because the config filename has changed, it makes sense to change the command line `--no-eslintrc` flag to a more generic name, `--no-config-file` and change `CLIEngine`'s `useEslintrc` option to `useConfigFile`. In the short term, to avoid a breaking change, these pairs of names can be aliased to each other.
+
+## Documentation
+
+This will require extensive documentation changes and an introductory blog post.
+
+At a minimum, these pages will have to be updated (and rewritten):
+
+* https://eslint.org/docs/user-guide/getting-started#configuration
+* https://eslint.org/docs/user-guide/configuring
+* https://eslint.org/docs/developer-guide/working-with-plugins#configs-in-plugins
+* https://eslint.org/docs/developer-guide/shareable-configs
+
+## Drawbacks
+
+As with any significant change, there are some significant drawbacks:
+
+1. We'd need to do a phased rollout (see Backwards Compatibility Analysis) to minimize impact on users. That means maintaining two configuration systems for a while, increasing maintenance overhead and complexity of the application.
```

Comment:
I'm not sure having a codemod is possible. Given that people could be using any of four different formats for their current config file, plus mixing in `extends` and `overrides`, I think it would be difficult to go in and modify those configs and make them work.

I initially thought we might be able to provide a function that could at least convert existing configs after they are read in, but once again, the complexities of merging, module resolving, etc., makes that appear impractical.

Overall, I don't think the format is different enough that it would be difficult for people to convert themselves.

Reply:
I'm not sure I understand the difficulty of doing this automatically. Could you clarify?

* It seems trivial for a tool to look at a project and determine which of the four config formats a project is using (it could just check for the existence of those files in the filesystem, like ESLint currently does).
* `extends: foo` can be transformed within a single config to put `require('eslint-config-foo')` in a config array.
* A similar transformation applies to `overrides`:

    ```js
    // old config
    module.exports = {
        ...baseConfig,
       overrides: {
           files: ["foo"],
           ...overrideConfig
       },
    };
    ```

    ```js
    // equivalent new config
    exports.config = [
        baseConfig,
        {
            files: ["foo"],
            ...overrideConfig
        }
    ];
    ```

* A runtime conversion tool also doesn't seem very complicated because we already have code that handles merging/module resolving (namely, the current config logic within ESLint).

I agree that the format isn't very difficult for a human to convert if they know what they're doing, but I think most of our users aren't very familiar with how configs work, and making everyone learn the new config format is a very large amount of migration pain.

Doing a back-of-the-envelope calculation:

* We have about 30 million downloads per week on npm. Assuming an average active project downloads ESLint about 30 times per week, there are about a million projects with old-style config files.
* Supposing it takes an average user 20 minutes to read our new config documentation and understand how to fix their config, that's about 333333 person-hours, or 38 person-years, spent migrating configs from the old format to the new format.

These are very rough estimates and could easily be off by a factor of four in either direction, but my point is that making our existing users learn our new config format would create a massive amount of work for them. Creating a tool to do the migration automatically would be a huge time-saver, and I think it's worth doing even if working out the details takes some time.

Reply:
@not-an-aardvark this is already handled by using `@eslint/config` utilities? I don't see anything more specific to do.